qiskit/circuit/quantumcircuit.py

```diff
@@ -1352,55 +1325,56 @@ def qasm(self, formatted=False, filename=None, encoding=None):
                 ``True``.
             QasmError: If circuit has free parameters.
         """
-        from qiskit.circuit.controlledgate import ControlledGate
 
         if self.num_parameters > 0:
             raise QasmError("Cannot represent circuits with unbound parameters in OpenQASM 2.")
 
         existing_gate_names = [
```

Comment:
This list should only include the names in of gates in `qelib1.inc` (p10-12 of https://arxiv.org/abs/1707.03429)

Reply:
Yes, but more gates definitions have been added to `qelib1.inc` inside qiskit code itself so all those gates can be parsed: https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/qasm/libs/qelib1.inc

Reply:
I just saw your new PR about reverting to the paper definition: https://github.com/Qiskit/qiskit-terra/pull/6125 I will change my code.

Reply:
`existing_gate_names` need to match with those defined here: https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/qasm/libs/qelib1.inc otherwise it may produce `Duplicate declaration` error. So my PR is dependent on yours: #6125

Reply:
Can I keep the extra gates definition for now and make another PR when this one #6125 gets merged?